Enable Neon optimizations within libwebrtc's Opus.

mstyura opened the pull request:
#### bc484ce61b8e9baa952030a708ab44d6d8a9867c
<pre>
Enable Neon optimizations within libwebrtc&apos;s Opus.

<a href="https://bugs.webkit.org/show_bug.cgi?id=295575">https://bugs.webkit.org/show_bug.cgi?id=295575</a>

Reviewed by Youenn Fablet.

Added missing files and preprocessor definitions.
Mostly mimic of what was done within BUILD.gn.

* Source/ThirdParty/libwebrtc/Configurations/opus.xcconfig:
* Source/ThirdParty/libwebrtc/libwebrtc.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/298080@main">https://commits.webkit.org/298080@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2eece926095ed421cd60f66ed0c8cc4180eb0d28

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/113160 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32895 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23373 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119368 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/64377 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33547 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41458 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86134 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/41467 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/116107 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26793 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101793 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66454 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26066 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19923 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63122 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96177 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19997 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122585 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40238 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30018 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/94982 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40622 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98009 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94724 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24310 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39867 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17673 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/36370 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40124 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39765 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43098 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41502 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Source/ThirdParty/libwebrtc/Configurations/opus.xcconfig:
```diff
@@ -13,16 +13,17 @@ INSTALL_PATH = $(INSTALL_PATH_PREFIX)$(WK_LIBRARY_INSTALL_PATH);
-GCC_PREPROCESSOR_DEFINITIONS = VAR_ARRAYS OPUS_BUILD OPUS_EXPORT= HAVE_LRINT HAVE_LRINTF $(inherited);
-GCC_PREPROCESSOR_DEFINITIONS[arch=arm64*] = $(inherited) OPUS_ARM_PRESUME_AARCH64_NEON_INTR;
+GCC_PREPROCESSOR_DEFINITIONS = USE_ALLOCA HAVE_ALLOCA_H OPUS_BUILD ENABLE_HARDENING OPUS_EXPORT= HAVE_LRINT HAVE_LRINTF $(inherited);
+GCC_PREPROCESSOR_DEFINITIONS[arch=arm64*] = $(inherited) OPUS_ARM_MAY_HAVE_NEON_INTR OPUS_ARM_PRESUME_NEON OPUS_ARM_PRESUME_NEON_INTR OPUS_ARM_PRESUME_AARCH64_NEON_INTR;
```
At least Opus 1.5.2 (which I realise is more recent than what we have vendored from libwebrtc)'s `autoconf` script just results in `OPUS_ARM_MAY_HAVE_NEON_INTR OPUS_ARM_PRESUME_NEON_INTR OPUS_ARM_PRESUME_AARCH64_NEON_INTR` (i.e., no `OPUS_ARM_PRESUME_NEON`, which seems only to be set for A32 and not A64).

mstyura replied:
This particular macro was required to enable expansion of [silk_NSQ_del_dec](https://github.com/WebKit/WebKit/blob/1b7e1237692895204401904a60cfe68e65df9079/Source/ThirdParty/libwebrtc/Source/third_party/opus/src/silk/float/wrappers_FLP.c#L164) to [silk_NSQ_del_dec_neon](https://github.com/WebKit/WebKit/blob/1b7e1237692895204401904a60cfe68e65df9079/Source/ThirdParty/libwebrtc/Source/third_party/opus/src/silk/arm/NSQ_del_dec_arm.h#L34). The macro expansion [is dependent](https://github.com/WebKit/WebKit/blob/34cdb37c365064c6c5254a59776a200f45bec02a/Source/ThirdParty/libwebrtc/Source/third_party/opus/src/silk/arm/NSQ_del_dec_arm.h#L54) on [PRESUME_NEON](https://github.com/WebKit/WebKit/blob/1322eea39d8f8bcaa4d562c9edca2f5016c7b6bc/Source/ThirdParty/libwebrtc/Source/third_party/opus/src/celt/arm/armcpu.h#L62) which only defined if `‎OPUS_ARM_PRESUME_NEON` is defined.
I also found that in [CMakeLists.txt](https://github.com/WebKit/WebKit/blob/1b7e1237692895204401904a60cfe68e65df9079/Source/ThirdParty/libwebrtc/Source/third_party/opus/src/CMakeLists.txt#L497) both macros were defined.

gsnedders commented on Source/ThirdParty/libwebrtc/Configurations/opus.xcconfig:
```diff
@@ -13,16 +13,17 @@ INSTALL_PATH = $(INSTALL_PATH_PREFIX)$(WK_LIBRARY_INSTALL_PATH);
 WARNING_CFLAGS = -Wno-incompatible-pointer-types $(inherited)
-GCC_PREPROCESSOR_DEFINITIONS = VAR_ARRAYS OPUS_BUILD OPUS_EXPORT= HAVE_LRINT HAVE_LRINTF $(inherited);
-GCC_PREPROCESSOR_DEFINITIONS[arch=arm64*] = $(inherited) OPUS_ARM_PRESUME_AARCH64_NEON_INTR;
+GCC_PREPROCESSOR_DEFINITIONS = USE_ALLOCA HAVE_ALLOCA_H OPUS_BUILD ENABLE_HARDENING OPUS_EXPORT= HAVE_LRINT HAVE_LRINTF $(inherited);
```
I'm curious about why the move from `VAR_ARRAYS` to `USE_ALLOCA` — I'd have presumed the (standardised) `VAR_ARRAYS`, which is also the upstream default, would make more sense?

mstyura replied:
I've took for reference what google did in [their BUILD.gn](https://github.com/WebKit/WebKit/blob/1b7e1237692895204401904a60cfe68e65df9079/Source/ThirdParty/libwebrtc/Source/third_party/opus/BUILD.gn#L34) for Opus. It's not required to enable neon, and maybe I should not have put irrelevant to a bug change here. Just let me know if you think it's better I undo this part.

mstyura replied:
I've read more about the reasoning behind Chromium change to  enable `USE_ALLOCA`, they refer to [article](https://www.phoronix.com/news/Linux-Kills-The-VLA) which tells that Linux kernel prohibited to use VLAs. But the Linux kernel has switched to use fixed size arrays, not `alloca`. So I've updated PR and removed changes unrelated to usage of `neon`.

gsnedders reviewed (commented):
This is really something for the other reviewers I added, but:

mstyura committed:
Enable Neon optimizations within libwebrtc's Opus.

https://bugs.webkit.org/show_bug.cgi?id=295575

Reviewed by Youenn Fablet.

Added missing files and preprocessor definitions.
Mostly mimic of what was done within BUILD.gn.

* Source/ThirdParty/libwebrtc/Configurations/opus.xcconfig:
* Source/ThirdParty/libwebrtc/libwebrtc.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/298080@main
